### DISCUSSION TITLE
Touchscreen gestures

### DISCUSSION THREAD

**Discussion by minego** [niri-wm/niri#463] [Ideas]
I've recently purchased a minisforum v3 tablet, and I'm very happy with how niri runs on it, but it is rather unusable in tablet mode, since I have no way to switch windows or workspaces.

I'd love to be able to bind support for gestures on the touchscreen, in a similar way to the trackpad. As a starting point I think the same gestures that the trackpad supports would be sufficient, but being able to do others like swiping from a screen edge, would be very useful as well.

I hope to dig into the source and try to see if I can make this work, but I am not sure where to start.

**Comment by YaLTeR** [comment]
Yeah, it would be good to have some touchpad gestures (and also finish missing touch input in the screenshot UI). However, I think this depends on gesture bindings, as I've got feedback to *not* add 3-finger touch swipes (at least, by default) because that prevents apps from using 3 finger touches themselves, which can be limiting (think a game console emulator where you need to press 3 buttons at once). For the gesture bindings, I don't have a design finalized, only some thoughts here: https://github.com/YaLTeR/niri/issues/372

**Comment by minego** [comment]
Personally I feel like applications wanting to use it is rare enough that I'd prefer to have it on by default... But maybe that is a more common use case than I thought? It feels to me like using gestures to navigate is more important than other use cases it may conflict with though. I mean, if I have to use a keyboard and mouse to start a game anyway...

Or, perhaps it would make sense to enable them by default, but disable them for full screen windows or based on app rules?

I don't know. Mostly just brain dumping now.

**Comment by minego** [comment]
I just had a thought... For my purposes, if I could tap a button on my bar to switch to a mode where swiping with 1 finger acts like the gestures, I'd be pretty happy... Not perfect, but a good start.

I could add a button in waybar to send a `niri msg` command to turn on the mode, then switch windows or workspaces, etc... Right now I have no way at all to do these things without the keyboard attached, since virtual keyboards can't trigger niri bindings.

I'm hoping to find some time to play with the niri source and see if I can experiment with this, but that will take a bit of learning (I don't know rust well...) and I don't know when I'll find the time.

**Comment by minego** [comment]
I spent a while looking at how the 3 finger touchpad gestures work, to try to see if I could make them work with a touchscreen... I was hoping I could add an action that can be triggered with `niri msg action` to turn on or off a mode where any pointer event would be treated like a 3 button gesture.

But, it doesn't seem to be as generic as that... I'm assuming that is because of how Smithay or libinput works.

I may make another attempt later...

**Reply by YaLTeR** [reply]
FWIW the same 3 finger gesture is currently hooked up to both the pointer input (Mod+MMB) and to the touchpad input. (With a slight difference to how the gesture works to account for the different input device.) For a touchscreen the deceleration multiplier will be slightly different, otherwise it'll be the same as touchpad.

This is all completely orthogonal to adding an action like this of course. But, I'm wondering if it should even be an action in the first place? Rather than just a gesture bind (which needs finishing the design for those).

**Reply by minego** [reply]
> FWIW the same 3 finger gesture is currently hooked up to both the pointer input (Mod+MMB) and to the touchpad input. (With a slight difference to how the gesture works to account for the different input device.) For a touchscreen the deceleration multiplier will be slightly different, otherwise it'll be the same as touchpad.
> 
> This is all completely orthogonal to adding an action like this of course. But, I'm wondering if it should even be an action in the first place? Rather than just a gesture bind (which needs finishing the design for those).

Oh, really? I'll have to dig more then! 

Yeah, an action may not be the right way to do this but, for a first step for experimenting it will be enough. Right now my plan is to have a button on my bar that can trigger the action, but long term some gesture would be much better. Perhaps something like pinch in with 4 or 5 fingers...

**Comment by itshog** [comment]
I'd really like to be able to use the same three fingers scrolling gestures currently implemented for touchpad input when using my convertible laptop in tablet mode: the scrolling paradigm Niri uses is IMO already a better option than "classic" tiling for tablets and convertibles, and having the option to scroll or change workspace with a three fingers swipe could be a massive improvement. We could finally have an intuitive, elegant and user friendly interface for convertible laptops on linux, without relying on dozens of buttons in the bar of the wm just for basic functionality.

I know about [lisgd](https://git.sr.ht/~mil/lisgd) (the program used for touch gestures on Sxmo if I'm not mistaken), and probably with some tweaking it could be made to work, but I believe that a native solution which mimics the already amazing touchpad gestures has some value. Also, I couldn't find in the output of `niri msg action` a command that works the same as the three finger horizontal swipe to use with lisgd (but maybe I just didn't notice it).

I apologize beforehand in case there are technical reasons preventing the implementation of touchscreen gestures, like missing functionality in Smithay or something else.

**Reply by YaLTeR** [reply]
With gesture bindings it should be possible to bind 3 finger touch swipe to the same thing as touchpad.

> Also, I couldn't find in the output of niri msg action a command that works the same as the three finger horizontal swipe

Well, yeah. Actions are discrete, whereas the touchpad swipes are continuous.

**Reply by itshog** [reply]
> With gesture bindings it should be possible to bind 3 finger touch swipe to the same thing as touchpad.

Looking forward to this! My experience of using tiling wms when in tablet mode has been a sort of nightmare (at least when talking about navigation between windows), this would be kind of a dream come true.
 
> Well, yeah. Actions are discrete, whereas the touchpad swipes are continuous.

Thanks for the clarification.

**Reply by YaLTeR** [reply]
That said, I wonder if there's a different touch gesture set that does not restrict the apps to 1-2 fingers. Maybe something closer to tablet OSes where swiping from the bottom brings up an overview where you can swipe and select a window with 1 finger?

**Reply by itshog** [reply]
> That said, I wonder if there's a different touch gesture set that does not restrict the apps to 1-2 fingers.

Personally I'd be fine with three (and more) fingers reserved to gestures and 1-2 reserved to applications (as this mimics the touchpad interaction), but I can see the problem here. A "silly" option would be to distinguish between gestures which initiates _inside_ a focused window, which could be forwarded to the applications and only if the app doesn't have a matching action be treated as gestures, and gestures which initiates _outside_ a window, like on the bar or on the wallpaper, which should always be treated as gestures regardless of the number of fingers. The issue here is that this would work only if the user has some "free space" on the desktop for using gestures which conflicts with applications (like by having gaps and/or a bar), but not all users have this "free space".

**Reply by YaLTeR** [reply]
> only if the app doesn't have a matching action

Well, this is a problem because apps don't tell the compositor whether or not they handled any input that came their way. And without this, using only the free space may be quite annoying I imagine.

But yeah, configurable 3 finger always eaten by compositor should suffice for now.

**Reply by itshog** [reply]
Another option would be requiring gestures to start from a screen edge like in Android with the "back",  "show all open apps" and "switch focused app" actions, but this doesn't completely solve the problem. For the record, the problem is present in Android too: for instance, if you have the web browser urlbar on the bottom of the screen, a right/left swipe on the urlbar should switch tab, but in lots of cases it ends up getting registered as a gesture and it switches application instead.

**Reply by YaLTeR** [reply]
Yeah, also apps should be able to use left and right edge swipes too I feel. But, I think bottom and top edge swipes can be reserved for the compositor (similarly to how they are on tablet/phone OSes). In GNOME, a top edge swipe is already reserved to unfullscreening, and bottom edge swipe to bringing up the on-screen keyboard if I'm not mistaken?

**Reply by itshog** [reply]
Maybe this could be left up to the user, by having separate `touch-swipe-3` and `touch-swipe-edge-3` options. With the first, the user knows that this may conflict with applications, while the second is a little bit less ergonomic on big screens but resembles more the way mobile operating systems usually work (and in this case it could be possible to have also `touch-swipe-edge-1` and `touch-swipe-edge-2`).

**Reply by itshog** [reply]
So this way we could combine both approach: gestures from anywhere on the screen starting with a minimum of 3 fingers, and edge gestures starting with a minimum of one.

**Reply by YaLTeR** [reply]
Will need to see yeah.

**Reply by orowith2os** [reply]
> Yeah, also apps should be able to use left and right edge swipes too I feel.

FWIW, on Android, Firefox can hijack left-right gestures from the pill bar to switch tabs. It would be kind of nice to have this be possible under Linux as well. I think that might involve a lot more communication on what capabilities an app has, and when they should be forwarded to the compositor instead. I know right now the compositor dictates that alone, without app input. So more Wayland protocol development.

**Reply by charnam** [reply]
@orowith2os If the goal is to prevent directional drags on the window from being intercepted by the compositor, by means of an app-based communication standard, then this protocol would very likely need to be supported on GNOME and KDE before seeing widespread adoption, which would _definitely_ all only occur after significant Wayland protocol development. This specific situation, therefore, probably isn't the place for the Niri project itself to be making decisions. If anyone's goal is to get a base app-to-compositor communication feature such as this one standardized, then the target is instead likely to be in the Wayland protocol bug tracker.

**Comment by inmymum** [comment]
I run niri on a large 17 inch laptop with a touchscreen so something like this would be legendary. What I imagine is something like how kde does it, which is 3 finger swipe to the sides to switch to the next workspace. I think a option in the config to enable something like this mode would be great. It is not enirely neccesary, however this would be a massive qol feature and might bring many users to niri if they have a touchscreen laptop like me.

**Comment by ethernet-gravy** [comment]
Is there any work planned towards this? As a convertible laptop user, I'm pretty interested in seeing some way to switch windows through the touchscreen itself. I'm willing to volunteer my time towards implementing this (although I might need some guidance).

**Comment by shimarulin** [comment]
My currently workaround for use Niri on touchscreen with [lisgd](https://git.sr.ht/~mil/lisgd):

```sh
lisgd -d /dev/input/by-id/usb-ELAN_Touchscreen-event-if00 -t 10 -T 5 -g "1,RL,R,*,R,niri msg action focus-column-right" -g "1,LR,L,*,R,niri msg action focus-column-left" -g "1,DU,B,*,R,niri msg action open-overview"
```

Three-finger gestures on the touchscreen don't seem very comfortable. But I tried to handle gestures from the edges of the screen with one finger and it looks good, but sometimes there is duplication of events.

How it works: 

https://github.com/user-attachments/assets/f2a5e437-f01b-4a20-965b-af94f321f38b

**Reply by ethernet-gravy** [reply]
Thanks a lot! Looks like just what I need.

**Reply by shimarulin** [reply]
I wrote a script that runs in parallel with [iio-niri](https://github.com/Zhaith-Izaliel/iio-niri) so that lisgd restarts when the screen orientation changes to correctly detect the edges of the screen. It also tries to automatically detect the available touch screen device. I tried evtest, but it works very slowly, so I settled on a simpler detection.

```sh
#!/usr/bin/env bash

# Function for automatic touch device detection
detect_touch_device() {
    echo "Searching for touch devices..." >&2

    # Search for devices by patterns, excluding stylus
    local patterns=(
        "/dev/input/by-id/*Touchscreen*event*"
        "/dev/input/by-id/*touch*event*"
    )

    for pattern in "${patterns[@]}"; do
        for device in $pattern; do
            # Exclude stylus devices by name
            if [[ "$device" != *"mouse"* ]] && \
               [[ "$device" != *"stylus"* ]] && \
               [[ "$device" != *"pen"* ]] && \
               [ -e "$device" ] && [ -r "$device" ]; then
                echo "Found device: $device" >&2
                echo "$device"
                return 0
            fi
        done
    done

    echo "Error: No touch devices found" >&2
    return 1
}

# Argument handling
if [ $# -eq 1 ]; then
    DEVICE="$1"

    # Check if device exists
    if [ ! -e "$DEVICE" ]; then
        echo "Error: Device $DEVICE does not exist"
        exit 1
    fi
    echo "Using specified device: $DEVICE"
elif [ $# -eq 0 ]; then
    echo "Auto-detecting touch device..."
    DEVICE=$(detect_touch_device)
    if [ $? -ne 0 ] || [ -z "$DEVICE" ]; then
        echo "Error: Failed to auto-detect compatible touch device"
        echo "Please specify device manually: $0 <device_path>"
        exit 1
    fi
    echo "Auto-detected compatible device: $DEVICE"
else
    echo "Usage: $0 [device_path]"
    echo "Example: $0 /dev/input/by-id/usb-ELAN_Touchscreen-event-if00"
    echo "If no device specified, will try to auto-detect compatible touch device"
    exit 1
fi

# Check device accessibility
if [ ! -r "$DEVICE" ]; then
    echo "Error: Device $DEVICE is not readable"
    echo "You may need to add your user to input group"
    exit 1
fi

# Function to stop all lisgd processes
stop_lisgd() {
    pkill -f "lisgd -d $DEVICE" 2>/dev/null
}

# Function to start/restart lisgd
start_lisgd() {
    # Stop previous instances
    stop_lisgd

    # Small delay before restart
    sleep 0.1

    # Start lisgd in background (&)
    echo "Starting lisgd with device: $DEVICE"
    lisgd -d "$DEVICE" -t 10 -T 5 \
        -g "1,RL,R,*,R,niri msg action focus-column-right" \
        -g "1,LR,L,*,R,niri msg action focus-column-left" \
        -g "1,DU,B,*,R,niri msg action open-overview" &

    # Save PID
    LISGD_PID=$!

    # Check if lisgd started successfully
    sleep 0.1
    if ! ps -p $LISGD_PID >/dev/null 2>&1; then
        echo "Error: lisgd failed to start. Check device permissions and compatibility."
        exit 1
    fi

    echo "lisgd started successfully with PID: $LISGD_PID"
}

# Stop lisgd when script exits
cleanup() {
    echo "Stopping lisgd..."
    stop_lisgd
    exit 0
}

trap cleanup EXIT INT TERM

# Initial startup
start_lisgd

# Monitor orientation changes
echo "Monitoring orientation changes..."
monitor-sensor 2>/dev/null | awk '/Accelerometer orientation changed:/ { print $NF; fflush(); }' | while read -r line
do
    case "$line" in
        normal|bottom-up|right-up|left-up)
            echo "Orientation changed: $line, restarting lisgd..."
            start_lisgd
            ;;
        *)
            echo "Unknown orientation: $line"
            ;;
    esac
done

# If monitor-sensor terminated
echo "Error: monitor-sensor terminated. Exiting."
cleanup

```

**Reply by Sempyos** [reply]
I found rot8 works lovely because it allows running/passing current rotation to a script

**Reply by shimarulin** [reply]
Do you rotate the display and touchscreen for Niri too using https://github.com/efernau/rot8 ? Like `rot8 --display eDP-1 --hooks ./lisgd-niri.sh`?

**Reply by Sempyos** [reply]
Yup.

**Reply by Sempyos** [reply]
Except mine needs the -XY arguments.

**Comment by oknozor** [comment]
Hey there, if looking for an alternative to lisgd I wrote a small daemon that allow to define callbacks for configurable touch screen gestures (swipe, long press, edge configuration)

https://github.com/oknozor/roland

**Example:**

```toml
[[gestures]]
num_fingers = 2
kind = "SwipeUp"
action = "notify-send 'Two Finger Swipe Up (Bottom Edge)'"
min_duration = 50
min_distance = 100.0
on_edge = { Bottom = 300 }

 [[gestures]]
 num_fingers = 3
 kind = "SwipeDown"
 action = "notify-send 'Three Finger Swipe Down'"
 min_duration = 50
 mix_distance = 100.0

[[gestures]]
num_fingers = 4
kind = "Press"
action = "notify-send 'Long Press' 'Four finger long press detected'"
min_duration = 500
max_distance = 20.0

 # etc ... 
```

```
roland --config config.toml
```

**Reply by Sempyos** [reply]
That's very cool !
Would there be more to it in the future or is it going to just stay as is .?

And how uh is it compared to lisgd..?

**Reply by oknozor** [reply]
Note that I wrote this for myself only because lisgd is lacking edge detection.
  
I hope @YaLTeR will be cooking something similar directly integrated in Niri in the future, so I am not planning to actively maintain it. But if people actually improve a lot of stuff. 

Below are the key differences:  

| Feature | Roland | lisgd |
|---------|--------|-------|
| **Primary Use Case** | Touch gestures for Niri (Wayland) | Touch gestures (libinput) |
| **Configuration** | TOML file | config.h or command-line flags |
| **Gesture Types** | Swipe (4 directions), Press | Swipe (8 directions including diagonals) |
| **Multi-finger Support** | Yes | Yes |
| **Edge Detection** | Yes (configurable per gesture) | No |
| **Duration Constraints** | Yes (min/max duration) | Yes (timeoutms) |
| **Distance Constraints** | Yes (min/max distance) | Yes (distancethreshold) |
| **Orientation Support** | No | Yes (0-3 rotations) |
| **Angle Leniency** | No | Yes (degreesleniency) |
| **Event Source** | libinput (udev) | libinput (device file) |

**Reply by inmymum** [reply]
> Note that I wrote this for myself only because lisgd is lacking edge detection.
> 
> I hope @YaLTeR will be cooking something similar directly integrated in Niri in the future, so I am not planning to actively maintain it. But if people actually improve a lot of stuff.
> 
> Below are the key differences:
> 
> Feature	Roland	lisgd
> **Primary Use Case**	Touch gestures for Niri (Wayland)	Touch gestures (libinput)
> **Configuration**	TOML file	config.h or command-line flags
> **Gesture Types**	Swipe (4 directions), Press	Swipe (8 directions including diagonals)
> **Multi-finger Support**	Yes	Yes
> **Edge Detection**	Yes (configurable per gesture)	No
> **Duration Constraints**	Yes (min/max duration)	Yes (timeoutms)
> **Distance Constraints**	Yes (min/max distance)	Yes (distancethreshold)
> **Orientation Support**	No	Yes (0-3 rotations)
> **Angle Leniency**	No	Yes (degreesleniency)
> **Event Source**	libinput (udev)	libinput (device file)

How efficient is it compared to lisgd? (As far as cpu usage as battery life is subjective to the laptop, and configuration)

**Reply by oknozor** [reply]
I did not ran any benchmark, so it's up to you to discover I guess :)

**Comment by charnam** [comment]
Touchscreen gestures could be implemented in this way:

Swiping left and right at the bottom of the screen (similar to modern smartphones) would horizontally scroll your workspace. Similarly, swiping up from the bottom should open the overview. Dragging a window with three (or optionally four, or optionally disabled) fingers could act as picking it up, similar to Mod+Drag. If this behavior is not desired, you could also make window titles draggable, for an idea. Ideally, things should be more configurable, but IMO getting at least something working would be good to do first.

I used to run Hyprland on my phone, but I switched to using Android for a year due to features not working. Developers have since made progress and I'm ready to try mobile Linux again, but I really do prefer Niri now. It would be nice to have Niri running on my phone, when the time comes to reinstall Linux on my phone.

I'd like if this was added as a native part of Niri, because it would be much less buggy/hacky than other solutions I've read about here. Of course, you are all volunteers, and I'm not expecting this, but it would be really, really cool.

**Reply by inmymum** [reply]
> Touchscreen gestures could be implemented in this way:
> 
> Swiping left and right at the bottom of the screen (similar to modern smartphones) would horizontally scroll your workspace. Similarly, swiping up from the bottom should open the overview. Dragging a window with three (or optionally four, or optionally disabled) fingers could act as picking it up, similar to Mod+Drag. If this behavior is not desired, you could also make window titles draggable, for an idea. Ideally, things should be more configurable, but IMO getting at least something working would be good to do first.
> 
> I used to run Hyprland on my phone, but I switched to using Android for a year due to features not working. Developers have since made progress and I'm ready to try mobile Linux again, but I really do prefer Niri now. It would be nice to have Niri running on my phone, when the time comes to reinstall Linux on my phone.
> 
> I'd like if this was added as a native part of Niri, because it would be much less buggy/hacky than other solutions I've read about here. Of course, you are all volunteers, and I'm not expecting this, but it would be really, really cool.

Additionally, if there was something like android swipe on edge to go back, it would be useful. Imagine if you could configure it on a per app basis to do different useful things. Just a thought.

**Reply by Sempyos** [reply]
you can checkout lisgd for the moment
https://git.sr.ht/~mil/lisgd

**Reply by inmymum** [reply]
Of course. They said it themselves quite well: "I'd like if this was added as a native part of Niri, because it would be much less buggy/hacky than other solutions I've read about here."

**Comment by pir0c0pter0** [comment]
### Working implementation of touchscreen 3-finger swipe gestures

I've implemented native touchscreen 3-finger swipe gesture support for workspace switching (vertical) and view scrolling (horizontal), and I'd like to get feedback on the approach before opening a PR.

**How it works:**

The implementation adds a `TouchSwipeGestureTracker` that collects raw `touch_down`/`touch_motion`/`touch_up` events across multiple slots, recognizes when 3 fingers are swiping in the same direction (using the same 16px threshold as touchpad gestures), and then feeds deltas into the existing `workspace_switch_gesture_*` / `view_offset_gesture_*` layout pipeline — the same one the touchpad uses. This means the touchscreen carousel gets the same continuous animation, inertia, and spring physics for free.

The gesture lifecycle is managed through a `TouchSwipeGrab` (modeled on the existing `TouchOverviewGrab`), with `is_touchpad: false` passed to the layout gesture API.

**Key implementation detail:** smithay creates an implicit touch grab when the first finger touches a surface, which prevents subsequent fingers from being seen by normal input handlers. The tracker is fed all touch events unconditionally (before grab checks) to work around this, and the swipe grab is only installed once the gesture is recognized.

**Configuration:**

```kdl
gestures {
    touchscreen-swipe {
        fingers 3
        // off
    }
}
```

Enabled by default. Can be disabled with `off`. The finger count is configurable. I'm aware this intersects with the gesture bindings design in #372 — the config structure is intentionally minimal and could be migrated into whatever binding system is eventually adopted.

**Regarding the app conflict concern:** @YaLTeR mentioned that 3-finger touches could prevent apps from receiving multi-touch input. In this implementation, the gesture is recognized from the raw touch events before they reach applications, same as how the touchpad 3-finger swipe works. The `off` flag provides an escape hatch. A future enhancement could add edge-only triggering as discussed above.

**Testing & verification:**
- Tested on a oneXPlayer tablet (GXTP7380 touchscreen, i2c HID) running CachyOS + niri 25.11
- Workspace switching (vertical swipe) and horizontal view scrolling both work with continuous finger-following animation
- `cargo test --all`: 203 tests pass, 0 failures
- `cargo +nightly fmt --all`: clean, no formatting issues
- 11 unit tests for the gesture tracker covering: recognition threshold, direction detection, centroid deltas, finger lift/cancel, extra fingers, reset

**Commits (3 clean, each builds and passes tests):**
1. `feat(input): add touchscreen 3-finger swipe gestures` — tracker, grab, input wiring
2. `feat(config): add touchscreen-swipe gesture configuration` — config types and parsing
3. `docs: document touchscreen swipe gestures` — wiki pages

**Branch:** https://github.com/pir0c0pter0/niri-touch-gestures/tree/feat/touchscreen-swipe-gestures

Happy to adjust the approach based on feedback before opening a formal PR.

**Reply by charnam** [reply]
_Edit: We can ignore the above post. It's completely AI generated. Still have to wait for an actual implementation._

Good progress! Though, at some point in the future, I would like to see the ability to add single-finger edge-swipe gestures.

Example:
- Swipe left and right on "bottom" **edge** to switch window in workspace (similar to modern smartphone gesture navigation)
- Swipe up from bottom **edge** to open overview
- Swipe down from top **edge** to `exec` something

Would this be possible already using your system? Using three fingers at once on my phone screen isn't easy when I'm holding it in one hand... Looking forward to your changes, though!

**Reply by lVentus** [reply]
> ### Working implementation of touchscreen 3-finger swipe gestures
> I've implemented native touchscreen 3-finger swipe gesture support for workspace switching (vertical) and view scrolling (horizontal), and I'd like to get feedback on the approach before opening a PR.我已经实现了原生触摸屏三指滑动操作对工作区切换（垂直）和视图滚动（水平）的支持，我想在提交 PR 之前获得一些关于此方法的反馈。
> 
> **How it works:  工作原理：**
> 
> The implementation adds a `TouchSwipeGestureTracker` that collects raw `touch_down`/`touch_motion`/`touch_up` events across multiple slots, recognizes when 3 fingers are swiping in the same direction (using the same 16px threshold as touchpad gestures), and then feeds deltas into the existing `workspace_switch_gesture_*` / `view_offset_gesture_*` layout pipeline — the same one the touchpad uses. This means the touchscreen carousel gets the same continuous animation, inertia, and spring physics for free.该实现添加了一个 `TouchSwipeGestureTracker` ，用于收集多个槽位中的原始 `touch_down` / `touch_motion` / `touch_up` 事件，识别三指何时朝同一方向滑动（使用与触控板手势相同的 16px 阈值），然后将增量值传递给现有的 `workspace_switch_gesture_*` / `view_offset_gesture_*` 布局管道——与触控板使用的管道相同。这意味着触摸屏轮播动画可以免费获得与触控板相同的连续动画、惯性和弹簧物理效果。
> 
> The gesture lifecycle is managed through a `TouchSwipeGrab` (modeled on the existing `TouchOverviewGrab`), with `is_touchpad: false` passed to the layout gesture API.手势生命周期通过 `TouchSwipeGrab` （以现有的 `TouchOverviewGrab` 为模型）进行管理，并将 `is_touchpad: false` 传递给布局手势 API。
> 
> **Key implementation detail:** smithay creates an implicit touch grab when the first finger touches a surface, which prevents subsequent fingers from being seen by normal input handlers. The tracker is fed all touch events unconditionally (before grab checks) to work around this, and the swipe grab is only installed once the gesture is recognized.**关键实现细节：** 当第一个手指触碰到表面时，smithay 会创建一个隐式触摸抓取事件，这会导致后续手指的触摸无法被正常的输入处理程序识别。为了解决这个问题，追踪器会无条件地接收所有触摸事件（在抓取检查之前），并且只有在识别到滑动手势后才会触发滑动抓取事件。
> 
> **Configuration:  配置：**
> 
> ```kdl
> gestures {
>     touchscreen-swipe {
>         fingers 3
>         // off
>     }
> }
> ```
> 
> Enabled by default. Can be disabled with `off`. The finger count is configurable. I'm aware this intersects with the gesture bindings design in #372 — the config structure is intentionally minimal and could be migrated into whatever binding system is eventually adopted.默认启用。可通过 `off` 禁用。手指数量可配置。我知道这与 #372 中的手势绑定设计有所冲突——配置结构有意设计得非常简洁，可以迁移到最终采用的任何绑定系统中。
> 
> **Regarding the app conflict concern:** @YaLTeR mentioned that 3-finger touches could prevent apps from receiving multi-touch input. In this implementation, the gesture is recognized from the raw touch events before they reach applications, same as how the touchpad 3-finger swipe works. The `off` flag provides an escape hatch. A future enhancement could add edge-only triggering as discussed above.**关于应用冲突问题：** 有人提到三指触摸可能会导致应用无法接收多点触控输入。在这个实现中，手势识别是在原始触摸事件到达应用之前进行的，这与触控板三指滑动的工作原理相同。 `off` 标志提供了一个备用方案。未来的改进可能会添加如上所述的边缘触发功能。
> 
> **Testing & verification:  测试与验证：**
> 
> * Tested on a oneXPlayer tablet (GXTP7380 touchscreen, i2c HID) running CachyOS + niri 25.11在运行 CachyOS + niri 25.11 的 oneXPlayer 平板电脑（GXTP7380 触摸屏，i2c HID）上进行了测试。
> * Workspace switching (vertical swipe) and horizontal view scrolling both work with continuous finger-following animation工作区切换（垂直滑动）和水平视图滚动都支持连续的手指跟随动画。
> * `cargo test --all`: 203 tests pass, 0 failures`cargo test --all` ：203 个测试通过，0 个测试失败
> * `cargo +nightly fmt --all`: clean, no formatting issues`cargo +nightly fmt --all` ：干净，无格式问题
> * 11 unit tests for the gesture tracker covering: recognition threshold, direction detection, centroid deltas, finger lift/cancel, extra fingers, reset针对手势追踪器，编写了 11 项单元测试，涵盖以下内容：识别阈值、方向检测、质心偏移、手指抬起/取消、额外手指、重置
> 
> **Commits (3 clean, each builds and passes tests):提交（3 次提交均已完成，每次提交都能构建并通过测试）：**
> 
> 1. `feat(input): add touchscreen 3-finger swipe gestures` — tracker, grab, input wiring `feat(input): add touchscreen 3-finger swipe gestures` — 跟踪器、抓取器、输入接线
> 2. `feat(config): add touchscreen-swipe gesture configuration` — config types and parsing `feat(config): add touchscreen-swipe gesture configuration` — 配置类型和解析
> 3. `docs: document touchscreen swipe gestures` — wiki pages `docs: document touchscreen swipe gestures` — 维基页面
> 
> **Branch:** https://github.com/pir0c0pter0/niri-touch-gestures/tree/feat/touchscreen-swipe-gestures**分支：** [https](https://github.com/pir0c0pter0/niri-touch-gestures/tree/feat/touchscreen-swipe-gestures) ://github.com/pir0c0pter0/niri-touch-gestures/tree/feat/touchscreen-swipe-gestures
> 
> Happy to adjust the approach based on feedback before opening a formal PR.我乐意根据反馈调整方法，然后再提交正式的 PR。

Is it possible to let user define how many fingers to use? I prefer to use 4 finger for touchpad and screen. 3 fingers for drag( double tap to drag is too stupid

**Reply by pir0c0pter0** [reply]
Hey @charnam, totally get that three fingers on a phone screen while holding it with one hand isn't great.

We looked into this — single-finger touch handles all normal interactions (tapping, dragging, selecting, moving windows), so just allowing `fingers 1` would make **any** 16px+ movement get captured as a gesture, making the touchscreen unusable for everything else.

What you're describing is an **edge swipe** — a gesture that only triggers when the touch starts near the screen edge. It's doable with relatively small changes. Here's a roadmap:

1. **Add an `edge_zone` field to the config** (`niri-config/src/gestures.rs`):
   - Add `edge_zone: u16` to `TouchscreenSwipe` (default `0` = disabled)
   - Add `edge_zone: Option<u16>` to `TouchscreenSwipePart`
   - Merge it in `MergeWith`

2. **Add validation in `merge_with`** — `fingers 1` should only be accepted when `edge_zone > 0`. `fingers 2` conflicts with two-finger scrolling. Valid values: `1` (with edge zone), `3`, `4`.

3. **Extend `TouchSwipeGestureTracker`** (`src/input/touch_swipe_tracker.rs`):
   - Add `edge_zone: f64` and `output_size: Option<(f64, f64)>` fields
   - Add a `set_output_size(width, height)` method
   - In `touch_down`, when `min_fingers == 1` and `edge_zone > 0`, reject touches that don't start within `edge_zone` pixels of any screen edge

4. **Wire up the output size** (`src/input/mod.rs`):
   - In `on_touch_down`, call `set_output_size()` on the tracker before feeding the touch event
   - In `on_touch_motion` where the tracker is recreated after a gesture, propagate the output size to the new tracker too

5. **Fix the hardcoded `3`** (`src/niri.rs` ~line 2521):
   - Read `fingers` and `edge_zone` from `config.borrow().gestures.touchscreen_swipe` instead of hardcoding `3`

The config would look like:
```
gestures {
    touchscreen-swipe {
        fingers 1
        edge-zone 20
    }
}
```

This keeps normal single-finger interactions working everywhere except the edge zone, which is exactly the smartphone-style navigation you're after.

**Reply by pir0c0pter0** [reply]
Hey @lVentus, the finger count is already configurable via the `fingers` field in the config. Niri's touchpad gestures use 3 fingers as their standard, so we followed the same convention for touchscreen and kept 3 as the default. But you can set it to 4 if you want:

```
gestures {
    touchscreen-swipe {
        fingers 4
    }
}
```

Just keep in mind this only affects **touchscreen** gestures — touchpad gestures have their own separate configuration and stay unchanged.

**Reply by charnam** [reply]
@pir0c0pter0 Please tell people you're an AI bot before wasting everyone's time, thanks.

**Comment by elouan660** [comment]
This would be really nice, this is actually the only missing feature that makes me stick to GNOME

**Comment by GGEZUS** [comment]
While waiting for [pull/3771](https://github.com/niri-wm/niri/pull/3771) to be integrated I created my own implementation of touch support. Check [niri-tablet](https://github.com/GGEZUS/niri-tablet) (based on [julianjc84](https://github.com/julianjc84/niri/tree/feat/configurable-touch-gestures) amazing work).
